driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java

```diff
@@ -19,203 +19,150 @@
-            if (isOperationInProgress) {
-                isClosePending = true;
-                return false;
-            } else if (!isClosed) {
-                isClosed = true;
-                isClosePending = false;
-                return true;
-            }
-            return false;
-        });
+        this.maxWireVersion = connectionDescription.getMaxWireVersion();
+        this.firstBatchEmpty = commandCursorResult.getResults().isEmpty();
 
-        if (doClose) {
-            killCursorOnClose();
+        AsyncConnection connectionToPin = null;
+        if (connectionDescription.getServerType() == ServerType.LOAD_BALANCER) {
+            connectionToPin = connection;
         }
+
+        resourceManager = new ResourceManager(namespace, connectionSource, connectionToPin, commandCursorResult.getServerCursor());
     }
 
     @Override
     public void next(final SingleResultCallback<List<T>> callback) {
         if (isClosed()) {
-            callback.onResult(null, new MongoException("next() called after the cursor was closed."));
-        } else if (initialCommandCursorResult != null && !firstBatchEmpty) {
-            // May be empty for a tailable cursor
-            List<T> results = initialCommandCursorResult.getResults();
-            initialCommandCursorResult = null;
-            if (getServerCursor() == null) {
-                close();
+            callback.onResult(null, new MongoException(MESSAGE_IF_CLOSED_AS_CURSOR));
+            return;
+        }
+
+        resourceManager.execute((AsyncCallbackSupplier<List<T>>) funcCallback -> {
+            ServerCursor localServerCursor = resourceManager.getServerCursor();
+            boolean serverCursorIsNull = localServerCursor == null;
+            List<T> batchResults = emptyList();
+            if (!processedInitial.getAndSet(true) && !firstBatchEmpty) {
+                batchResults = commandCursorResult.getResults();
             }
-            callback.onResult(results, null);
-        } else {
-            ServerCursor localCursor = getServerCursor();
-            if (localCursor == null) {
-                close();
-                callback.onResult(null, null);
+
+            if (serverCursorIsNull || !batchResults.isEmpty()) {
+                funcCallback.onResult(batchResults, null);
             } else {
-                boolean doGetMore = withLock(lock, () ->  {
-                    if (isClosed()) {
-                        callback.onResult(null, new MongoException("next() called after the cursor was closed."));
-                        return false;
-                    }
-                    isOperationInProgress = true;
-                    return true;
-                });
-                if (doGetMore) {
-                    getMore(localCursor, callback);
-                }
+                getMore(localServerCursor, funcCallback);
             }
-        }
+        }).get((r, t) -> {
+            if (resourceManager.getServerCursor() == null) {
```

Comment:
I think we can remove the auto-closing nature of these AsyncBatchCursors now.

I propose adding a `hasNext` method which just then returns `resourceManager.getServerCursor() != null` for async.

Reply:
I commented on what I think is the same change in the sync version, where it's easier to understand.

Reply:
This is different to the sync issue (or I missed which one was commented on). AsyncBatchCursors automatically mark themselves as closed on consumption of all resources. That predates these refactorings.

Reply:
I think my comment on the sync version is about something different, so ignore my previous comment on this thread.

Reply:
I _think_ what you're saying is that users of the batch cursors will now have to call `close` explicitly in order to release resources (i.e the pinned connection and the connection source).  If so, that is a behavioral change that we have to consider carefully, as it effects not just driver internals but users of the driver.  Consider what we've allowed in the legacy API:

```
DBCursor cursor = dbCollection.find();
while (cursor.hasNext()) {
   cursor.next();
}
// allow cursor to go out of scope and let the cursor cleaner call DBCursor#close, which in turn calls CommandBatchCursor#close
```

Currently, if the above loop exits normally (without throwing), all the resources will be released and the cursor cleaner is a no-op.  With this proposed change, the releases won't be released until GC kicks in.

While the above is shoddy application coding, we have allowed it and I'm hesitant to change its behavior unless there is a compelling reason.

Reply:
Ah that makes more sense, let me capture that in a test and apply the fix to release.

Reply:
nm - there are tests that already cover this scenario for example: 
> should handle getMore when there are empty results but there is a cursor

So there is no change in behaviour here - because setting the server cursor to null `resourceManger.setServerCursor` already releases the resources automatically if the cursor is null.

Reply:
I agree, that the code
```java
if (getServerCursor() == null) {
    close();
}
```
is not needed for releasing resources (and there is no such code in `CommandBatchCursor`), but it is needed to self-close the `AsyncCommandCatchCursor`. As far as I understand, without this code the `AsyncCommandCatchCursor.isClosed` will continue returning `false` despite all resources having been released (`CommandBatchCursor` does not have this issue because it does not have the `isClosed` method).

Thus, I think we should either live the code as is, and add a comment explaining what's going on.

Reply:
Even better, instead of doing
```
if (getServerCursor() == null) {
    close();
}
```
after `endOperation` in `AsyncCommandBatchCursor`, let's just call `close` instead of `releaseClientResources` in `CursorResourceManager.setServerCursor`. This will have the same effect, and `AsyncCommandBatchCursor` won't need the aforementioned additional code that we don't have in `CommandBatchCursor`.